### PR TITLE
Change the format of access tokens away from macaroons

### PR DESCRIPTION
**Pull request by richvdh**
Some history might be helpful here:

Access tokens were originally turned into macaroons back in 2015 (in #229 and #256); however, it has always been the case that *as well as* satisfying the caveats on the macaroons, the macaroon has to be stored in the database.

The fact that the access token has to be in the database meant that much of the power of macaroons was unusable (in particular, the ability to add third-party caveats), and much of the security they offer was redundant.
  
The spec, as far as I can recall, never required that access tokens be macaroons; matrix-org/matrix-doc#1204 discussed changing this but [concluded](https://docs.google.com/document/d/1mdis1LQcoOSVRElszEmrAWZGIX0jX_croSha-X5oe_w/edit#heading=h.eoyizjahtsqy) that they should be left specified as opaque strings.

#4374 changed the behaviour of synapse again so that the macaroons were never validated (though the token still had to be in the database). The logic behind that change was somewhat tied up in CVE-2019-5885 (#4664).

Once #4374 landed, the macaroonness of the access tokens was completely redundant, and they could equally well be any cryptographically-secure random string. This PR therefore brings us back full circle.

The reasons for switching away from macaroons include:

 * they are over 300 bytes long, which is a significant overhead when they have
   to be attached to every single request.
 * generating macaroons (in particular, signing them) costs valuable CPU cycles.
 * it's generally confusing.

----

The format used in this PR changed a bit as we worked on it. We ended up with:

`syt_<base64_localpart>_<20 random characters>_<base62 crc>`

... mostly inspired by https://github.blog/2021-04-05-behind-githubs-new-authentication-token-formats.

Whereas access tokens were around 260 bytes before (depending on the length of your localpart and domain), they will now be around 45 bytes (depending on the length of your localpart).

### CODE REVIEW

**Comment by richvdh** [comment]
I'm going to shelve this for now. It turns out we have some lua in our haproxy config which relies on access tokens looking like macaroons. I'm not sure the advantages of this change are worth the work.

**Comment by richvdh** [comment]
>  It turns out we have some lua in our haproxy config which relies on access tokens looking like macaroons.

this is no longer true, so we could revisit this I think.

**Comment by erikjohnston** [comment]
> > It turns out we have some lua in our haproxy config which relies on access tokens looking like macaroons.
> 
> this is no longer true, so we could revisit this I think.

I wonder if we should still encode the user ID in the tokens somehow, in a way that haproxy can efficiently read? A couple of options here are:

* Base64 encode the user ID (or localpart) then prepend the random string (after a deliminator). Then in haproxy we can use the `field` extractor to get something we can load balance on, and `b64dec` to get the actual user ID.
* Hash the user ID, base64, truncate to ten bytes, then add the random string after. In haproxy we can then simply load balance on the first ten bytes (though we wouldn't be able to get the user ID).

**Comment by richvdh** [comment]
yeah, that would probably make some sense.

Another thought is that we should probably take the opportunity to do something like https://github.blog/2021-04-05-behind-githubs-new-authentication-token-formats/.

**Comment by erikjohnston** [comment]
(I've taken this out of the review queue as Rich is planning on changing the format to incorporate some of the ideas above).

For a concrete suggestion: I think something like: `syt_<base64 local part>_<random string>` would work, it a) has a prefix to make it easily detectable and b) encodes the user local part in a way that can be extracted by haproxy (and anything else that can split by fields).

**Review by erikjohnston — COMMENTED** [review]
I think this LGTM, modulo anything that Dan's RFC throws up.

Do we want to run this past the ops team to make sure we're not missing anything they'd like? (While we're here)

File: synapse/handlers/auth.py [hunk 1]
```diff
@@ -1192,9 +1197,18 @@ async def _check_local_password(self, user_id: str, password: str) -> Optional[s
...
-        return stringutils.random_string(20)
+
+        # we use the following format for access tokens:
+        #    syt_<base64 local part>_<random string>_<base62 crc check>
+
+        b64local = unpaddedbase64.encode_base64(for_user.localpart.encode("utf-8"))
+        random_string = stringutils.random_string(20)
+        base = f"syt_{b64local}_{random_string}"
+
+        crc = base62_encode(crc32(base.encode("ascii")), minwidth=6)
```
[the remark this answers is not in the document]
Comment (richvdh) [reply synapse/handlers/auth.py:1210]: cargo-culted from https://github.blog/2021-04-05-behind-githubs-new-authentication-token-formats/#checksum

[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/handlers/auth.py:1210]: wfm

**Comment by callahad** [comment]
The GitHub / Slack / Stripe consensus seems to be:

- Prefixes which encode company + token type (e.g., `ghp` vs `gho` for GitHub Personal access tokens vs GitHub OAuth access token)
- Separating the prefix with a `_` to avoid accidental collisions with randomly generated strings (SHAs, etc.)

GitHub _also_ uses the last 6 characters of their tokens to embed a base62-encoded CRC32 checksum of the token

Their tokens are 40 characters long: prefix (3) + `_` + token (30) + checksum (6). The token itself is case-sensitive alphanumeric. 30 characters of that yields 178 bits of entropy.

File: synapse/handlers/auth.py [hunk 2]
```diff
@@ -1192,6 +1197,19 @@ async def _check_local_password(self, user_id: str, password: str) -> Optional[s
             return None
         return user_id
 
+    def generate_access_token(self, for_user: UserID) -> str:
+        """Generates an opaque string, for use as an access token"""
+
+        # we use the following format for access tokens:
+        #    syt_<base64 local part>_<random string>_<base62 crc check>
+
+        b64local = unpaddedbase64.encode_base64(for_user.localpart.encode("utf-8"))
```
[the remark this answers is not in the document]
Comment (richvdh) [reply synapse/handlers/auth.py:1206]: of course, we could just say that only people with [spec-compliant](https://matrix.org/docs/spec/appendices#common-identifier-format) mxids are allowed to use matrix, so it doesn't need encoding at all :smiling_imp:

**Comment by richvdh** [comment]
>  30 characters of that yields 178 bits of entropy.

for comparison: our 20 characters from a 52-character range gives 6.5*20 = 130 bits of entropy, though the attacker also has to match it up with the right localpart.

File: synapse/handlers/auth.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (richvdh) [reply synapse/handlers/auth.py:1206]: I think that's fair. it's really not that hard to do it right, even from a shell command.

**Comment by callahad** [comment]
I didn't actually see an approving review happen here? Did I miss it?

**Comment by babolivier** [comment]
@erikjohnston said:

> I think this LGTM, modulo anything that Dan's RFC throws up.

Though he didn't actually tick "approve".

**Comment by richvdh** [comment]
what @babolivier said. I didn't consider it worth disturbing people again since everyone seemed to be in agreement. Sorry if that was presumptuous.

**Comment by callahad** [comment]
Had this gone back into the review queue, I think I would've raised the following queries:

- The inclusion of `b64local` makes our tokens variable length, are we OK with that?
- The base62 vocabulary we're using is in a different order than just base64 without the final two characters, are we OK with that? (b64 uses `A-Za-z0-9+/`, our b62 uses `0-9A-Za-z`)
- Are we sure we're comfortable baking user identifiers into the token itself?
    - Is that a privacy risk? (e.g., inadvertently logging identifiable information)
    - Is that a DoS risk? (since similar local parts will encode similarly, and we'd presumably use these to direct traffic to workers per Erik's comment?)

**Comment by callahad** [comment]
If I'm reading the code correctly, the regex to match a Synapse Access Token is:

    syt_[a-zA-Z0-9+/]+_[a-zA-Z]{20}_[a-zA-Z0-9]{6}

The regex to match a GitHub Personal Access Token is 

    ghp_[a-zA-Z0-9]{36,}

...is the complexity in our format warranted?

**Comment by richvdh** [comment]
> Had this gone back into the review queue, I think I would've raised the following queries:

Most of these came up in online discussion as well. Thanks for raising them again, not least as a reminder to me to record the conclusions.

> * The inclusion of `b64local` makes our tokens variable length, are we OK with that?

They were variable length before and it was never a problem...

> * The base62 vocabulary we're using is in a different order than just base64 without the final two characters, are we OK with that? (b64 uses `A-Za-z0-9+/`, our b62 uses `0-9A-Za-z`)

I couldn't find any sort of convention for b62 alphabets, but most implementations I found with googling used that alphabet. I don't think it matters too much.

> * We're using _unpadded_ base64, but haproxy's `b64dec` expects proper padding, are we OK with that?

hrrrm, good question. I'd forgotten to address that. I think you can go from unpadded to padded base64 with a couple of string manipulations. Erik said "Don't that is a huge concern though, since I think conversion is relatively easy with simple string replacement?", though I'm not sure how you would actually do so, 

> * Are we sure we're comfortable baking user identifiers into the token itself?
>   * Is that a privacy risk? (e.g., inadvertently logging identifiable information)

Again, we've always done it before. I think if you're logging access tokens, you have bigger problems than the identifiable information.

>   * Is that a DoS risk? (since similar local parts will encode similarly, and we'd presumably use these to direct traffic to workers per Erik's comment?)

I'm not quite sure what sort of attack you're envisaging here. Similar local parts *will* encode similarly, but likely we would use that as a hash key for a sticky routing table, rather than forcing particular local parts to route to particular backends.

> Also, isn't the entropy 114 bits, not 130? `math.log2(52**20)`

errrr yes. `log2(52**20) = 20 * log2(52) =~ 20 * 5.7 = 114`. I don't know where I got 6.5 from before.


> If I'm reading the code correctly, the regex to match a Synapse Access Token is:
> ```
> syt_[a-zA-Z0-9+/]+_[a-zA-Z]{20}_[a-zA-Z0-9]{6}
> ```

Or possibly `syt_[a-zA-Z0-9+/_]+`. What's your objective here?

**Comment by callahad** [comment]
> > * We're using _unpadded_ base64, but haproxy's `b64dec` expects proper padding, are we OK with that?
> 
> hrrrm, good question. I'd forgotten to address that. I think you can go from unpadded to padded base64 with a couple of string manipulations. Erik said "Don't that is a huge concern though, since I think conversion is relatively easy with simple string replacement?", though I'm not sure how you would actually do so,

I understood Erik's comment to mean translating between urlsafe and normal base64 "with a couple of string manipulations", in the context of haproxy <2.4 not supportting the urlsafe variant

> > If I'm reading the code correctly, the regex to match a Synapse Access Token is:
> > ```
> > syt_[a-zA-Z0-9+/]+_[a-zA-Z]{20}_[a-zA-Z0-9]{6}
> > ```
> 
> Or possibly `syt_[a-zA-Z0-9+/_]+`. What's your objective here?

To unambiguously identify access tokens, e.g., to participate in programs like GitHub's [Secret Scanning](https://docs.github.com/en/developers/overview/secret-scanning) or to implement our own audits

**Comment by erikjohnston** [comment]
> > > * We're using _unpadded_ base64, but haproxy's `b64dec` expects proper padding, are we OK with that?
> > 
> > 
> > hrrrm, good question. I'd forgotten to address that. I think you can go from unpadded to padded base64 with a couple of string manipulations. Erik said "Don't that is a huge concern though, since I think conversion is relatively easy with simple string replacement?", though I'm not sure how you would actually do so,
> 
> I understood Erik's comment to mean translating between urlsafe and normal base64 "with a couple of string manipulations", in the context of haproxy <2.4 not supporting the urlsafe variant

FWIW the main use case is to route requests, where you don't need to decode the user ID and can just route based on the second field of the access token. 

> > > If I'm reading the code correctly, the regex to match a Synapse Access Token is:
> > > ```
> > > syt_[a-zA-Z0-9+/]+_[a-zA-Z]{20}_[a-zA-Z0-9]{6}
> > > ```
> > 
> > 
> > Or possibly `syt_[a-zA-Z0-9+/_]+`. What's your objective here?
> 
> To unambiguously identify access tokens, e.g., to participate in programs like GitHub's [Secret Scanning](https://docs.github.com/en/developers/overview/secret-scanning) or to implement our own audits

If the goal is to provide a regex that is going to have the least false positives then having a more complicated regex will *help* that? It's going to be very rare for anyone to want to parse one of these tokens, so I don't think the complexity of the regex (in terms of number of characters) is a huge concern here.

**Comment by callahad** [comment]
Works for me 👍

**Comment by erikjohnston** [comment]
(Those were all good Qs though)

**Comment by richvdh** [comment]
> FWIW the main use case is to route requests, where you don't need to decode the user ID and can just route based on the second field of the access token.

in which case maybe we should just have used a hash and saved ourselves the bother of worrying about whether it is ok to encode the localpart in the access token :innocent: 

However, I'm very much inclined to stick with what we have rather than agonise about it too much. (Though yes, all good questions worth discussing)

**Pull request merged**